### PR TITLE
Create Min.java

### PR DESCRIPTION
**Pull request by thkim4u**

### CODE REVIEW

**Review by vincent-kwon — APPROVED** [review]

File: Min.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      if (result == null)
+        throw new NullPointerException("Min.min");
+
+      while (itr.hasNext()) {   // throws NPE, CCE as needed
+        T comp = itr.next();
+        if (comp.compareTo(result) < 0) {
+            result = comp;
+        }
+      }
+      return result;
```
Comment (thkim4u) [inline Min.java:37]: Please, check indentation.

File: Min.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        throw new IllegalArgumentException("Min.min");
+      }
+      Iterator<? extends T> itr = list.iterator();
+      T result = itr.next();
+
+      if (result == null)
+        throw new NullPointerException("Min.min");
+
+      while (itr.hasNext()) {   // throws NPE, CCE as needed
+        T comp = itr.next();
```
Comment (thkim4u) [inline Min.java:32]: Null check seems to be placed here too like:
if (comp == null) {
        throw new NullPointerException("Min.min");
}

File: Min.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    * @throws ClassCastException
+    *         if list elements are not mutually comparable
+    * @throws IllegalArgumentException if list is empty
+    */
+  public static <T extends Comparable<? super T>>
+      T min (List<? extends T> list) {
+      if (list.size() == 0) {
+        throw new IllegalArgumentException("Min.min");
+      }
+      Iterator<? extends T> itr = list.iterator();
```
Comment (thkim4u) [inline Min.java:25]: Can you explain this line, expecially "Iterator<? extends T>"?

File: Min.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    * Returns the mininum element in a list
+    * @param list Comparable list of elements to search
+    * @return the minimum element in the list
+    * @throws NullPointerException if list is null or
+    *         if any list elements are null
+    * @throws ClassCastException
+    *         if list elements are not mutually comparable
+    * @throws IllegalArgumentException if list is empty
+    */
+  public static <T extends Comparable<? super T>>
```
Comment (thkim4u) [inline Min.java:20]: For this line, I refere the link: https://stackoverflow.com/questions/25779184/explanation-of-generic-t-extends-comparable-super-t-in-collection-sort-com

File: Min.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      T min (List<? extends T> list) {
+      if (list.size() == 0) {
+        throw new IllegalArgumentException("Min.min");
+      }
+      Iterator<? extends T> itr = list.iterator();
+      T result = itr.next();
+
+      if (result == null)
+        throw new NullPointerException("Min.min");
```
Comment (thkim4u) [inline Min.java:29]: How about rewrite the above lines like the following:
<pre><code>
if (itr.hasNext()) {
    T result = itr.next();
    if (result == null) {
        throw new NullPointerException("Min.min");
    }
}
</code></pre>

**Pull request merged**